pkg/agent/controller/networkpolicy/fqdn.go

```diff
@@ -0,0 +1,776 @@
+		return "", nil, 0, fmt.Errorf("invalid DNS message")
+	}
+	fqdn := strings.ToLower(msg.Question[0].Name)
+	lowestTTL := uint32(math.MaxUint32) // a TTL must exist in the RRs
+	responseIPs := map[string]net.IP{}
+	for _, ans := range msg.Answer {
+		switch r := ans.(type) {
+		case *dns.A:
+			if f.ofClient.IsIPv4Enabled() {
+				responseIPs[r.A.String()] = r.A
+				if r.Header().Ttl < lowestTTL {
+					lowestTTL = r.Header().Ttl
+				}
+			}
+		case *dns.AAAA:
+			if f.ofClient.IsIPv6Enabled() {
+				responseIPs[r.AAAA.String()] = r.AAAA
+				if r.Header().Ttl < lowestTTL {
+					lowestTTL = r.Header().Ttl
+				}
+			}
+		}
+	}
+	if len(responseIPs) > 0 {
+		klog.V(4).InfoS("Received DNS Packet with valid Answer", "IPs", responseIPs, "TTL", lowestTTL)
+	}
+	if strings.HasSuffix(fqdn, ".") {
+		fqdn = fqdn[:len(fqdn)-1]
+	}
+	return fqdn, responseIPs, lowestTTL, nil
+}
+
+func (f *fqdnController) worker() {
+	for f.processNextWorkItem() {
+	}
+}
+
+func (f *fqdnController) processNextWorkItem() bool {
+	key, quit := f.dnsQueryQueue.Get()
+	if quit {
+		return false
+	}
+	defer f.dnsQueryQueue.Done(key)
+
+	err := f.makeDNSRequest(key.(string))
+	f.handleErr(err, key)
+	return true
+}
+
+func (f *fqdnController) handleErr(err error, key interface{}) {
+	if err == nil {
+		f.dnsQueryQueue.Forget(key)
+		return
+	}
+	klog.ErrorS(err, "Error syncing FQDN, retrying", "fqdn", key)
+	f.dnsQueryQueue.AddRateLimited(key)
+}
+
+// makeDNSRequest makes a proactive query for a FQDN to the coreDNS service.
+func (f *fqdnController) makeDNSRequest(fqdn string) error {
+	if f.dnsServerAddr == "" {
+		return fmt.Errorf("no DNS server is ready for query")
+	}
+	klog.InfoS("Making DNS request", "fqdn", fqdn, "dnsServer", f.dnsServerAddr)
+	dnsClient := dns.Client{SingleInflight: true}
+	fqdnToQuery := fqdn
+	// The FQDN in the DNS request needs to end by a dot
+	if fqdn[len(fqdn)-1] != '.' {
+		fqdnToQuery = fqdn + "."
+	}
+	query := func(m *dns.Msg) (*dns.Msg, error) {
+		r, _, err := dnsClient.Exchange(m, f.dnsServerAddr)
+		if err != nil {
+			klog.ErrorS(err, "DNS exchange failed")
+			return nil, err
+		}
+		return r, nil
+	}
+	v4ok, v6ok := true, true
+	if f.ofClient.IsIPv4Enabled() {
+		m := dns.Msg{}
+		m.SetQuestion(fqdnToQuery, dns.TypeA)
+		lookupTime := time.Now()
+		if res, err := query(&m); err == nil {
+			f.onDNSResponseMsg(res, lookupTime, nil)
+		} else {
+			v4ok = false
+		}
+	}
+	if f.ofClient.IsIPv6Enabled() {
+		m := dns.Msg{}
+		m.SetQuestion(fqdnToQuery, dns.TypeAAAA)
+		lookupTime := time.Now()
+		if res, err := query(&m); err == nil {
+			f.onDNSResponseMsg(res, lookupTime, nil)
+		} else {
+			v6ok = false
+		}
+	}
+	if !v4ok || !v6ok {
+		return fmt.Errorf("DNS request failed for at least one of type A or AAAA queries")
+	}
+	return nil
+}
+
+// implements openflow.PacketInHandler
+func (f *fqdnController) HandlePacketIn(pktIn *ofctrl.PacketIn) error {
+	if pktIn == nil {
+		return errors.New("empty packetin for Antrea Policy")
+	}
```

Comment:
I don't think we should have this check. There is something seriously wrong with the code if `pktIn` can be nil here